workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/TextController.kt

```diff
@@ -0,0 +1,166 @@
+   * render pass would trigger NxM useless render passes.
+   *
+   * However, when subscribing from the view layer, we _want_ the initial value to ensure the
+   * view gets initialized. We only want to drop the initial value when subscribing from
+   * a workflow, which is the only intended use case for [onTextChanged]. Because the text
+   * view wired up by [control], we don't actually need to expose the stateful nature of this
+   * flow anywhere.
+   *
+   * Instead, only text _change_ events are exposed, as [onTextChanged], which is suitable for use as a
+   * worker. The current value is exposed as a separate var, [textValue].
+   */
+  private val _textValue: MutableStateFlow<String> = MutableStateFlow(initialValue)
+
+  /**
+   * A [Flow] that emits the text value whenever it changes -- and only when it changes, the current value
+   * is not provided at subscription time. Workflows can safely observe changes by
+   * converting this value to a worker. (When using multiple instances, remember to provide unique
+   * key values to each `asWorker` call.)
+   *
+   * If you can do processing that doesn't require running a `WorkflowAction` or triggering a render
+   * pass, it can be done in regular Flow operators before converting to a worker.
+   */
+  public val onTextChanged: Flow<String> =
+  // Drop the first value because the StateFlow will always immediately emit the current value on
+    // subscription.
+    _textValue.drop(1)
+
+  /**
+   * The current text value.
+   */
+  public var textValue: String
+    get() = _textValue.value
+    set(value) {
+      _textValue.value = value
+    }
+
+  /**
+   * Call this from your LayoutRunner's showRendering method. This method is idempotent: if it has
+   * already been called with a particular [EditText], and the view has not been detached since the
+   * last call, it will do nothing. If a different [TextController]'s [control] is called on the same
+   * [EditText], the old one will be disconnected and the new one will replace it.
+   *
+   * See [TextController] for more documentation.
+   */
+  public fun control(view: EditText) {
+    // Do nothing if already subscribed on a previous update pass and the coroutine is still active.
+    val registeredController =
+      view.getTag(R.id.text_controller_rendering) as? TextControllerSubscription
+    if (registeredController?.controller === this &&
+      // This check ensures the subscription is re-started if the view was somehow detached since the
+      // call, eg. in a RecyclerView.
+      registeredController.subscription.isActive
+    ) {
+      return
+    }
+
+    // If we're replacing a controller with a different one, cancel the previous subscription.
+    registeredController?.subscription?.cancel()
+
+    view.setText(_textValue.value)
+    val subscription = view.launchWhenAttached {
+      _textValue
+        .onEach { textValue ->
```

Comment:
```suggestion
        .distinctUntilChanged()
        .onEach { textValue ->
```

I guess this necessitates another round of testing, but we might eke out a bit of optimization/debouncing on this side this way?

Reply:
That's already built into `MutableSharedFlow`.